frappe/model/base_document.py

```diff
@@ -83,14 +83,24 @@ def get_controller(doctype):
 
 	return site_controllers[doctype]
 
+def get_cached_doctype_data(doctype):
+	if not hasattr(frappe.local, "doctype_cache"):
+		frappe.local.doctype_cache = {}  # Initialize if not present
+
+	if doctype not in frappe.local.doctype_cache:
+		# Fetch and cache the data only if not already stored
+		frappe.local.doctype_cache[doctype] = frappe.db.get_value(
+			"DocType", doctype, ("module", "custom", "is_tree"), as_dict=True)
+
+	return frappe.local.doctype_cache[doctype]
 
 def import_controller(doctype):
 	from frappe.model.document import Document
 	from frappe.utils.nestedset import NestedSet
 
 	module_name = "Core"
 	if doctype not in DOCTYPES_FOR_DOCTYPE:
-		doctype_info = frappe.db.get_value("DocType", doctype, ("module", "custom", "is_tree"), as_dict=True)
+		doctype_info = get_cached_doctype_data(doctype)
```

Comment:
This is not required, the output of the entire function is cached. Check line 82